spec/features/sign_up_spec.rb

```diff
@@ -0,0 +1,108 @@
+require 'spec_helper'
+
+feature 'sign up form' do
+
+  scenario "can return to homepage" do
+    visit('/users/new')
+    click_button('Back')
+    expect(page).to have_current_path('/')
+  end
+
+  context 'password verification' do
+    context 'password matches' do
+      scenario 'can sign up' do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password')
+        expect(current_path).to eq('/peeps')
+        expect(page).to have_content("Welcome, test@test.com")
+      end
+
+      scenario 'signing up increases user count' do
+        expect{sign_up('name', 'user', 'test@test.com', 'password', 'password')}.to change{User.count}.by(1)
+      end
+
+      scenario "is empty" do
+        expect{sign_up('name', 'user', 'test@test.com', nil, 'password')}.to change{User.count}.by(0)
+      end
+    end
+
+    context "mismatched password" do
+      scenario "doesn't add a user" do
+        expect{sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')}.to change{User.count}.by(0)
+      end
+
+      scenario "is empty" do
+        expect{sign_up('name', 'user', 'test@test.com', 'password', nil)}.to change{User.count}.by(0)
+      end
+
+      scenario "leaves user on /users page" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(current_path).to eq('/users/new')
+      end
+
+      scenario "throws an error message" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(page).to have_content("Password does not match the confirmation")
+      end
+
+      scenario "shows original email address in sign up form" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(find_field('email').value).to eq('test@test.com')
+      end
+
+      scenario "shows original name in sign up form" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(find_field('name').value).to eq('name')
+      end
+
+      scenario "shows original username in sign up form" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(find_field('username').value).to eq('user')
+      end
```

Comment:
These are very thorough specs, but they'll take a while to run because each time they sign up a new user. If lots of specs are related, it's ok to have multiple expectations about related things.